dozer-sink-aerospike/src/lib.rs

```diff
@@ -0,0 +1,983 @@
+    #[error("Could not create record")]
+    CreateRecordError,
+}
+
+#[derive(Debug, Error)]
+struct AerospikeError {
+    code: i32,
+    message: String,
+}
+
+impl From<as_error> for AerospikeError {
+    fn from(value: as_error) -> Self {
+        let code = value.code;
+        let message = unsafe {
+            let message = CStr::from_ptr(value.message.as_ptr());
+            // The message is ASCII (I think?), so this should not fail
+            message.to_str().unwrap()
+        };
+        Self {
+            code,
+            message: message.to_owned(),
+        }
+    }
+}
+
+impl Display for AerospikeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} - {}", self.code, self.message)
+    }
+}
+
+// Client should never be `Clone`, because of the custom Drop impl
+#[derive(Debug)]
+struct Client {
+    inner: NonNull<aerospike>,
+}
+
+// The aerospike client API is thread-safe.
+unsafe impl Send for Client {}
+unsafe impl Sync for Client {}
+
+macro_rules! t {
+    ($e:expr, $err:path) => {{
+        let res = $e;
+        if res == as_status_e_AEROSPIKE_OK {
+            Ok(())
+        } else {
+            Err($err.assume_init().into())
+        }
+    }};
+}
+
+impl Client {
+    fn new(hosts: &str) -> Result<Self, AerospikeError> {
+        let mut config = unsafe {
+            let mut config = MaybeUninit::uninit();
+            as_config_init(config.as_mut_ptr());
+            config.assume_init()
+        };
+        unsafe {
+            // The hosts string should never contain inner null bytes, and if
+            // it does, the user is doing something very weird.
+            let hosts = CString::new(hosts).unwrap();
+            // The hosts string will be copied, so pass it as `as_ptr` so the original
+            // gets deallocated at the end of this block
+            as_config_add_hosts(&mut config as *mut as_config, hosts.as_ptr(), 3000);
+        }
+        // Allocate a new client instance. Our `Drop` implementation will make
+        // sure it is destroyed
+        let this = unsafe {
+            let inner = aerospike_new(&mut config as *mut as_config);
+            if inner.is_null() {
+                handle_alloc_error(Layout::new::<aerospike>())
+            } else {
+                let this = Self {
+                    inner: NonNull::new_unchecked(inner),
+                };
+                this.connect()?;
+                this
+            }
+        };
+        Ok(this)
+    }
+
+    unsafe fn connect(&self) -> Result<(), AerospikeError> {
+        unsafe {
+            let mut err = MaybeUninit::uninit();
+            t!(
+                aerospike_connect(self.inner.as_ptr(), err.as_mut_ptr()),
+                err
+            )
+        }
+    }
+
+    unsafe fn put(
+        &self,
+        key: *const as_key,
+        record: *mut as_record,
+        policy: as_policy_write,
+    ) -> Result<(), AerospikeError> {
+        unsafe {
```

Comment:
We don't need to put `unsafe` in `unsafe fn`?

Reply:
Yup. I do think in general it's a good idea to put the actually unsafe parts into separate unsafe blocks, each with their own `SAFETY:` comment explaining why they're sound, but with a single-line function, that's of course not very useful, and those comments don't currently exist :P